bob-apps/bin/bobp.rs

```diff
@@ -254,92 +267,199 @@ impl Debug for TaskConfig {
     }
 }
 
-type CodeRepresentation = i32; // Because tonic::Code does not implement hash
-
-#[derive(Default)]
-struct Statistics {
-    put_total: AtomicU64,
-    put_error_count: AtomicU64,
+struct PeriodicStatisticsBit {
+    count_spd: u64,
+    error: u64,
+    latency: f64,
+}
 
-    get_total: AtomicU64,
-    get_error_count: AtomicU64,
+struct PeriodicOperationStatistics<'a> {
+    count: DiffContainer<'a, u64>,
+    time_st: DiffContainer<'a, u64>,
+    count_st: DiffContainer<'a, u64>,
+    period_ms: u64,
+    stat: &'a OperationStatistics
+}
 
-    exist_total: AtomicU64,
-    exist_error_count: AtomicU64,
+impl<'a> PeriodicOperationStatistics<'a> {
+    fn from_operation_statistics(os: &'a OperationStatistics, period_ms: u64) -> Self {
+        Self {
+            count: DiffContainer::new(Box::new(|| os.total.load(Ordering::Relaxed))),
+            time_st: DiffContainer::new(Box::new(|| os.time_ns_st.load(Ordering::Relaxed))),
+            count_st: DiffContainer::new(Box::new(|| os.count_st.load(Ordering::Relaxed))),
+            period_ms,
+            stat: os,
+        }
+    }
 
-    put_time_ns_st: AtomicU64,
-    put_count_st: AtomicU64,
+    fn current_bit(&mut self) -> PeriodicStatisticsBit {
+        let diff = self.count.get_diff();
+        let count_spd = diff * 1000 / self.period_ms;
+        let cur_st_time = self.time_st.get_diff() as f64;
+        let cur_st_count = self.count_st.get_diff() as f64;
+        let error = self.stat.error_count.load(Ordering::Relaxed);
+        let latency = finite_or_default(cur_st_time / cur_st_count / 1e9);
+        PeriodicStatisticsBit {
+            count_spd,
+            error,
+            latency,
+        }
+    }
+}
 
-    get_time_ns_st: AtomicU64,
-    get_count_st: AtomicU64,
+struct PeriodicPutStatistics<'a> {
+    common: PeriodicOperationStatistics<'a>,
+    k: f64,
+}
 
-    exist_time_ns_st: AtomicU64,
-    exist_count_st: AtomicU64,
+impl<'a> PeriodicPutStatistics<'a> {
+    fn from(ps: &'a PutStatistics, period_ms: u64, k: f64) -> Self {
+        Self {
+            common: PeriodicOperationStatistics::from_operation_statistics(&ps.common, period_ms),
+            k,
+        }
+    }
 
-    get_errors: Mutex<HashMap<CodeRepresentation, u64>>,
-    put_errors: Mutex<HashMap<CodeRepresentation, u64>>,
-    exist_errors: Mutex<HashMap<CodeRepresentation, u64>>,
+    fn process_and_print(&mut self) -> f64 {
+        let bit = self.common.current_bit();
+        let spd = bit.count_spd as f64 * self.k;
+        println!("put:   {:>6} rps | err {:5} | {:>6.2} kb/s | lat {:>6.2} ms", 
+            bit.count_spd, bit.error, spd, bit.latency);
+        spd
+    }
+}
 
-    get_size_bytes: AtomicU64,
+struct PeriodicDeleteStatistics<'a> {
+    common: PeriodicOperationStatistics<'a>,
+    k: f64,
```

Comment:
`k` is a bad name. From the original code it is `request_size_kb`

Reply:
Also, receive this argument in `from` function in bytes (because originally it is in bytes) and then convert it to kb